src/internal/storage/fileset/validator.go

```diff
@@ -0,0 +1,71 @@
+package fileset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+)
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator validates the paths in a file set.
+func NewPathValidator(fs FileSet) FileSet {
+	return &pathValidator{
+		fs: fs,
+	}
+}
+
+func (pv *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	if err := pv.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr := checkCollisions(prev, idx); retErr != nil {
```

Comment:
This should be reverted back to `CheckIndex`, we want to perform the full validation. Any test that expects to be able to read an errored commit will need to be updated appropriately.